gen_breaktest.go

```diff
@@ -0,0 +1,215 @@
+// When/if the package is upgraded to a new version, change these to generate new tests.
+const (
+	url      = `https://www.unicode.org/Public/13.0.0/ucd/auxiliary/GraphemeBreakTest.txt`
+	filename = `GraphemeBreakTest-13.0.0.txt`
+)
+
+func main() {
+	log.SetPrefix("gen_breaktest: ")
+	log.SetFlags(0)
+
+	// Read text of testcases and parse into Go source code.
+	src, err := readAndParse()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Format the Go code.
+	srcfmt, err := format.Source(src)
+	if err != nil {
+		log.Fatalln("gofmt:", err)
+		//srcfmt = src
+	}
+
+	// Write it out.
+	if err := ioutil.WriteFile("grapheme_break_test.go", srcfmt, 0644); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// readAndParse reads a GraphemeBreakTest text file,
+// either from a local file or from a URL.
+//
+// It parses the file data into Go source code representing the testcases.
+func readAndParse() ([]byte, error) {
+	var r io.ReadCloser
+	if f, err := os.Open(filename); err == nil {
+		log.Printf("using %q", filename)
+		r = f
+	} else if !errors.Is(err, os.ErrNotExist) {
+		return nil, err
+	} else {
+		log.Printf("using %q", url)
+		resp, err := http.Get(url)
+		if err != nil {
+			return nil, err
+		}
+		r = resp.Body
+	}
+	defer r.Close()
+
+	buf := new(bytes.Buffer)
+	buf.Grow(120 << 10)
+	buf.WriteString(`// Code generated by go generate from gen_breaktest.go. DO NOT EDIT.
+
+package uniseg
+
+// unicodeTestCases are Grapheme testcases taken from
+// ` + url + `,
+// see https://www.unicode.org/license.html for the Unicode license agreement.
+var unicodeTestCases = []testcase {
+`)
+
+	sc := bufio.NewScanner(r)
+	num := 1
+	var line []byte
+	if sc.Scan() {
+		// Check first line for "# filename"
+		line = sc.Bytes()
+		if len(line) != 2+len(filename) || !strings.HasSuffix(string(line), filename) {
+			return nil, fmt.Errorf(`line %d: exected "# %v", got %q`, num, filename, line)
+		}
+	}
+
+	original := make([]byte, 0, 64)
+	expected := make([]byte, 0, 64)
+	for sc.Scan() {
+		num++
+		line = sc.Bytes()
+		if len(line) == 0 || line[0] == '#' {
+			continue
+		}
+		var comment []byte
+		if i := bytes.IndexByte(line, '#'); i >= 0 {
+			comment = bytes.TrimSpace(line[i+1:])
+			line = bytes.TrimSpace(line[:i])
+		}
+		original, expected, err := parseRuneSequence(line, original[:0], expected[:0])
+		if err != nil {
+			return nil, fmt.Errorf(`line %d: %v: %q`, num, err, line)
+		}
+		fmt.Fprintf(buf, "\t{original: \"%s\", expected: %s}, // %s\n", original, expected, comment)
+	}
+	if err := sc.Err(); err != nil {
+		return nil, err
+	}
+
+	// Check for final "# EOF", useful check if we're streaming via HTTP
+	if !bytes.Equal(line, []byte("# EOF")) {
+		return nil, fmt.Errorf(`line %d: exected "# EOF" as final line, got %q`, num, line)
+	}
+	buf.WriteString("}\n")
+	log.Printf("processed %d lines to %d/%d bytes", num, buf.Len(), buf.Cap())
+	return buf.Bytes(), nil
+}
+
+// used by parseRuneSequence to match input via bytes.HasPrefix
```

Comment:
Just a minor thing: Can you start this sentence with a capital "U" and end it with a period "."?

Reply:
Done.